crates/bevy_reflect/src/type_registry.rs

```diff
@@ -350,11 +350,103 @@ impl<T: for<'a> Deserialize<'a> + Reflect> FromType<T> for ReflectDeserialize {
     }
 }
 
+#[derive(Clone)]
+pub struct ReflectFromPtr {
+    type_id: TypeId,
+    to_reflect: for<'a> unsafe fn(*const (), lt: PhantomData<&'a ()>) -> &'a dyn Reflect,
+    to_reflect_mut: for<'a> unsafe fn(*mut (), lt: PhantomData<&'a ()>) -> &'a mut dyn Reflect,
+}
+
+impl ReflectFromPtr {
+    /// # Safety:
+    /// - `val` must be a pointer to a value of the type that the [`ReflectFromPtr`] was constructed for
+    /// - the lifetime `'a` of the return type can be arbitrarily chosen by the caller, who must ensure that during
+    /// that lifetime, `val` is valid
+    pub unsafe fn to_reflect_ptr<'a>(&self, val: *const ()) -> &'a dyn Reflect {
+        (self.to_reflect)(val, PhantomData)
+    }
+
+    /// # Safety:
+    /// - `val` must be a pointer to a value of the type that the [`ReflectFromPtr`] was constructed for
+    /// - the lifetime `'a` of the return type can be arbitrarily chosen by the caller, who must ensure that during
+    /// that lifetime, `val` is valid and not aliased
+    pub unsafe fn to_reflect_ptr_mut<'a>(&self, val: *mut ()) -> &'a mut dyn Reflect {
+        (self.to_reflect_mut)(val, PhantomData)
+    }
+
+    pub fn to_reflect<'a, T: 'static>(&self, val: &'a T) -> &'a dyn Reflect {
+        assert_eq!(self.type_id, std::any::TypeId::of::<T>());
+        // SAFE: the lifetime of `val` is the same as the lifetime of the return value
+        // and the type of `val` has been checked to be the same correct one
+        unsafe { self.to_reflect_ptr(val as *const _ as *const ()) }
+    }
+
+    pub fn to_reflect_mut<'a, T: 'static>(&self, val: &'a mut T) -> &'a mut dyn Reflect {
+        assert_eq!(self.type_id, std::any::TypeId::of::<T>());
+        // SAFE: the lifetime of `val` is the same as the lifetime of the return value
+        // and the type of `val` has been checked to be the same correct one
+        unsafe { self.to_reflect_ptr_mut(val as *mut _ as *mut ()) }
+    }
+}
+
+impl<T: Reflect + 'static> FromType<T> for ReflectFromPtr {
```

Comment:
Another nit: I believe `Reflect` already has the `'static` bound (unless you're adding it as a guard in case that changes).

```suggestion
impl<T: Reflect> FromType<T> for ReflectFromPtr {
```